## Paired T-testing: For Hayes Roth dataset b/w accuracies of (Scratch implementation & Scikit implementation)

In [24]:
from scipy import stats

knn_from_scratch_accuracy = [38.46153846153847, 38.46153846153847, 23.076923076923077, 38.46153846153847, 46.15384615384615, 61.53846153846154, 38.46153846153847, 23.076923076923077, 30.76923076923077, 61.53846153846154]
knn_scikit_accuracy = [38.46153846153847, 38.46153846153847, 23.076923076923077, 38.46153846153847, 46.15384615384615, 61.53846153846154, 38.46153846153847, 23.076923076923077, 30.76923076923077, 61.53846153846154]

# Perform a paired T-test
t_stat, p_value = stats.ttest_rel(knn_from_scratch_accuracy, knn_scikit_accuracy)

# Print the results
print("T-statistic:", t_stat)
print("P-value:", p_value)

# Decide whether to reject the null hypothesis
alpha = 0.05  # Significance level
if p_value < alpha:
    print("Reject the null hypothesis: There is a significant difference in accuracies.")
else:
    print("Fail to reject the null hypothesis: There is no significant difference in accuracies.")


T-statistic: nan
P-value: nan
Fail to reject the null hypothesis: There is no significant difference in accuracies.


## Paired T-testing: For Car_evaluation dataset

In [25]:
from scipy import stats

knn_from_scratch_accuracy = [65.69767441860465, 64.53488372093024, 65.69767441860465, 70.93023255813954, 61.627906976744185, 63.95348837209303, 66.27906976744185, 65.11627906976744, 58.139534883720934, 62.7906976744186]
knn_scikit_accuracy = [75.14450867, 74.56647399, 68.78612717, 75.14450867, 71.0982659,  72.25433526, 77.4566474,  68.20809249, 73.8372093,  69.76744186]

# Perform a paired T-test
t_stat, p_value = stats.ttest_rel(knn_from_scratch_accuracy, knn_scikit_accuracy)

# Print the results
print("T-statistic: ", t_stat)
print("P-value:", p_value)

# Decide whether to reject the null hypothesis
alpha = 0.05  # Significance level
if p_value < alpha:
    print("Reject the null hypothesis: There is a significant difference in accuracies.")
else:
    print("Fail to reject the null hypothesis: There is no significant difference in accuracies.")


T-statistic:  -6.506570037043797
P-value: 0.00011060277139652063
Reject the null hypothesis: There is a significant difference in accuracies.


## Paired T-testing: For Breast_Cancer dataset

In [26]:
from scipy import stats

knn_from_scratch_accuracy = [78.57142857142857, 78.57142857142857, 75.0, 64.28571428571429, 82.14285714285714, 82.14285714285714, 57.14285714285714, 78.57142857142857, 75.0, 67.85714285714286]
knn_scikit_accuracy = [79.31034483, 75.86206897, 58.62068966, 75.86206897, 75.86206897, 79.31034483, 85.71428571, 67.85714286, 67.85714286, 82.14285714]

# Perform a paired T-test
t_stat, p_value = stats.ttest_rel(knn_from_scratch_accuracy, knn_scikit_accuracy)

# Print the results
print("T-statistic: ", t_stat)
print("P-value:", p_value)

# Decide whether to reject the null hypothesis
alpha = 0.05  # Significance level
if p_value < alpha:
    print("Reject the null hypothesis: There is a significant difference in accuracies.")
else:
    print("Fail to reject the null hypothesis: There is no significant difference in accuracies.")


T-statistic:  -0.21377761645359086
P-value: 0.8354850043838716
Fail to reject the null hypothesis: There is no significant difference in accuracies.


## Conclusion
### -> Hence, we can see that only for Car_Evaluation Dataset Null Hypothesis(H0) can be rejected since the p-value is significantly lower than 0.05(i.e. 0.00011060277139652063) and T-statistic value that means the "Differences of their means between both accuracy arrays" is significantly higher than the other accuracy (i.e. -6.506570037043797). Therefore, I can say that only for Car_Evaluation dataset KNN_Using_Scikit implementation has better accuracies than the Scratch python implementation, one possible reason is that for this dataset only 1728 data rows are there unlike other two datasets.

### -> For remaining two datasets Breast_Cancer and Hayes_Roth since Null Hypothesis(H0) has FAILED to get rejected, both implementations of KNN are almost similar by looking at P-value and T-value we can identify that thing as well.

### -> More details will be found in the Word report.

## References

1. https://www.analyticsvidhya.com/blog/2021/09/hypothesis-testing-in-machine-learning-everything-you-need-to-know/
2. https://towardsdatascience.com/hypothesis-testing-in-machine-learning-using-python-a0dc89e169ce
3. https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_ind.html
4. https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_rel.html